packages/attune/src/bin/attune/cmd/apt/pkg/remove.rs

```diff
@@ -0,0 +1,176 @@
+    /// Distribution to remove the package from
+    #[arg(long, short)]
+    distribution: String,
+    /// Component to remove the package from
+    #[arg(long, short)]
+    component: String,
+
+    /// GPG key ID to sign the index with (see `gpg --list-secret-keys`)
+    #[arg(long, short)]
+    key_id: String,
+
+    /// Name of the package to remove
+    #[arg(long, short)]
+    package: String,
+    /// Version of the package to remove
+    #[arg(long, short)]
+    version: String,
+    /// Architecture of the package to remove
+    #[arg(long, short)]
+    architecture: String,
+}
+
+pub async fn run(ctx: Config, command: PkgRemoveCommand) -> ExitCode {
+    // Remove the package from the index, retrying if needed.
+    debug!("removing package from index");
+    let generate_index_request = GenerateIndexRequest {
+        change: PackageChange {
+            repository: command.repo.clone(),
+            distribution: command.distribution.clone(),
+            component: command.component.clone(),
+            action: PackageChangeAction::Remove {
+                name: command.package.clone(),
+                version: command.version.clone(),
+                architecture: command.architecture.clone(),
+            },
+        },
+    };
+    let res = ctx
+        .client
+        .get(
+            ctx.endpoint
+                .join(
+                    format!(
+                        "/api/v0/repositories/{}/index",
+                        percent_encode(command.repo.as_bytes(), PATH_SEGMENT_PERCENT_ENCODE_SET)
+                    )
+                    .as_str(),
+                )
+                .unwrap(),
+        )
+        .json(&generate_index_request)
+        .send()
+        .await
+        .expect("Could not send API request");
+    let (index, release_ts) = match res.status() {
+        StatusCode::OK => {
+            let res = res
+                .json::<GenerateIndexResponse>()
+                .await
+                .expect("Could not parse response");
+            debug!(index = ?res.release, "generated index to sign");
+            (res.release, res.release_ts)
+        }
+        status => {
+            let body = res.text().await.expect("Could not read response");
+            debug!(?body, ?status, "error response");
+            let error = serde_json::from_str::<ErrorResponse>(&body)
+                .expect("Could not parse error response");
+            eprintln!("Error removing package from index: {}", error.message);
+            return ExitCode::FAILURE;
+        }
+    };
+
+    // Sign index locally.
+    debug!(?index, "signing index");
+    let mut gpg = Context::from_protocol(Protocol::OpenPgp).expect("could not create gpg context");
+    gpg.set_armor(true);
+    let key = gpg
+        .find_secret_keys(vec![command.key_id.as_str()])
+        .expect("could not find secret key")
+        .next()
+        .expect("could not find secret key")
+        .expect("could not find secret key");
```

Comment:
[nit] Aren't these really all distinct error states? Ideally they'd have unique messages for less pain in debugging later.

Reply:
They are, but doesn't `expect` give a line number to disambiguate since it delegates to `panic!`?

Reply:
True, fair enough